src/conversations/conversations.service.ts

```diff
@@ -1,22 +1,99 @@
-import { Injectable } from '@nestjs/common';
+import { BadRequestException, Injectable } from '@nestjs/common';
 import { Conversation } from './entities/conversation.entity';
 import { Repository } from 'typeorm';
 import { InjectRepository } from '@nestjs/typeorm';
+import { HttpAuthGuard } from 'src/auth/guard/auth.guard';
+import { UsersService } from 'src/users/users.service';
+import { error } from 'console';
+import { UNAVAILABLE_GROUP, UNAVAILABLE_USER, UNAVAILABLE_USER_IN_CONVERSATION } from 'src/constant/error.constant';
+import { throwError } from 'rxjs';
+import { forEach } from 'lodash';
+import { createConversationID } from 'src/utils/ids';
+import { ConversationType } from 'src/constant/constant';
+import { type } from 'os';
+import { User } from 'src/users/entities/user.entity';
 
 @Injectable()
 export class ConversationsService {
   constructor(
     @InjectRepository(Conversation)
     private readonly conversationRepository: Repository<Conversation>,
+    private readonly userService : UsersService,
   ) {}
-  // async create(id : string) {
-  //   const newRoom = this.conversationRepository.create({
-  //     id
-  //   });
-  //   return this.conversationRepository.save(newRoom);
-  // }
-  // async find(id :string){
-  //   const chatroomFind = this.conversationRepository.findOneBy({id:id})
-  //   return chatroomFind;
-  // }
+
+  async findConversation(page : number, size : number, userId : number){
+    return await this.conversationRepository.createQueryBuilder("conversation")
+    .innerJoinAndSelect("conversation.users" , "user")
+    .leftJoinAndSelect("user.messages","message")
+    .select()
+    .where("\"userId\" = :userId",{userId : userId})
+    .take(size)
+    .skip((page - 1) * size)
+    .getMany()
+  }
+
+  async findOne(conversationId : string){
+    return await this.conversationRepository.createQueryBuilder("conversation")
+    .innerJoinAndSelect("conversation.users" , "user")
+    .leftJoinAndSelect("user.messages","message")
+    .select()
+    .where("\"conversationId\" = :id",{id : conversationId})
+    .getOne()
+  }
+
+  async createConversation(userCreateId : number ,groupName : string ,userIds : number[]){
+    const newConversationId = createConversationID();
+    const type = userIds.length >= 2 ? ConversationType.Group : ConversationType.Private;
+    const userCreate = await this.userService.getUserById(userCreateId)
+    const users: User[] =[JSON.parse(JSON.stringify(userCreate))];
+    for (let i = 0; i < userIds.length; i ++){
```

Comment:
// Sửa lại đoạn này thành như này
`
const getUserPromises = userIds.map(userId => this.userService.getUserById(userId));
users.concat(await Promise.all(getUserPromises));
`
// Làm cách cũ thì sẽ phải đợi get từng thằng user lần lượt, làm như này những cái get sẽ chạy concurrent